src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -390,6 +390,135 @@ BOOST_AUTO_TEST_CASE(SimpleLinearLayerTest)
   BOOST_REQUIRE_EQUAL(arma::accu(delta), 0);
 }
 
+/**
+ * Linear module weight initialization test.
+ */
+BOOST_AUTO_TEST_CASE(LinearLayerWeightInitializationTest)
+{
+  size_t inSize = 10, outSize = 4;
+  Linear<>* linear = new Linear<>(inSize, outSize);
+  linear->Reset();
+  RandomInitialization().Initialize(linear->Weight(), outSize, inSize);
+  linear->Bias().ones();
+
+  BOOST_CHECK_EQUAL_COLLECTIONS(linear->Weight().begin(),
+      linear->Weight().end(), linear->Parameters().begin(),
+      linear->Parameters().end() - outSize);
+
+  BOOST_CHECK_EQUAL_COLLECTIONS(linear->Bias().begin(),
+      linear->Bias().end(), linear->Parameters().begin() + inSize * outSize,
+      linear->Parameters().end());
+
+  BOOST_REQUIRE_EQUAL(linear->Weight().n_rows, outSize);
+  BOOST_REQUIRE_EQUAL(linear->Weight().n_cols, inSize);
+  BOOST_REQUIRE_EQUAL(linear->Bias().n_rows, outSize);
+  BOOST_REQUIRE_EQUAL(linear->Bias().n_cols, 1);
+  BOOST_REQUIRE_EQUAL(linear->Parameters().n_rows, inSize * outSize + outSize);
+
+  delete linear;
+}
+
+/**
+ * Atrous Convolution module weight initialization test.
+ */
+BOOST_AUTO_TEST_CASE(AtrousConvolutionLayerWeightInitializationTest)
+{
+  size_t inSize = 2, outSize = 3;
+  size_t kernelWidth = 4, kernelHeight = 5;
+  AtrousConvolution<>* module = new AtrousConvolution<>(inSize, outSize,
+      kernelWidth, kernelHeight, 6, 7, std::make_tuple(8, 9),
+      std::make_tuple(10, 11), 12, 13, 14, 15);
+  module->Reset();
+  RandomInitialization().Initialize(module->Weight(), kernelWidth, kernelHeight,
+      inSize * outSize);
+  module->Bias().ones();
```

Comment:
@lozhnikov: I don't know why the `module->Parameters()` contains the same values even after we initialize `module->Weight()` and `module->Bias()`. Though we can see from [here](https://github.com/mlpack/mlpack/blob/22e7c8e7d24c38c6837a0486fd0a45b5e2d7900c/src/mlpack/methods/ann/layer/atrous_convolution_impl.hpp#L151-L170) that we are not using copy to create `weight` and `bias`. But the same thing works for `Linear` layer.

Reply:
That's weird. I'll look into it. I need to build the project on my local computer.

Reply:
On further checking, I found that only the `weight` part is not modified. Change in `bias` is reflected in `module->Parameters()`.
I don't think there is any problem with `RandomInitialization` class.

I tried this and it works normal.
```cpp
  arma::mat weights;
  weights.set_size(10, 1);
  arma::cube weight = arma::cube(weights.memptr(), 2, 2, 2, false, false);
  arma::mat bias = arma::mat(weights.memptr() + weight.n_elem, 2, 1, false, false);
  RandomInitialization().Initialize(weight, 2, 2, 2);
  bias.ones();
  
  weight.print("weight:");
  weights.print("parameters:");
```

Reply:
It happens since `RandomInitialization` creates a new object. Indeed, look at the code:
```c++
  template<typename eT>
  void Initialize(arma::Cube<eT>& W,
                  const size_t rows,
                  const size_t cols,
                  const size_t slices)
  {
    // This line creates a new object.
    W = arma::Cube<eT>(rows, cols, slices);

    for (size_t i = 0; i < slices; i++)
      Initialize(W.slice(i), rows, cols);
  }
```
Just replace `RandomInitialization().Initialize()` by `model->Weight().randu()` and it'll work fine.

Reply:
@zoq Marcus, this might be interesting for you. Why didn't you write the function like the `arma::Mat` specialization
```c++
W = lowerBound + arma::randu<arma::Cube<eT>>(rows, cols, slices) *
        (upperBound - lowerBound);
```
Perhaps in this case Armadillo won't create a new object provided that there's enough memory.

Reply:
Maybe we can do something like this--
```cpp
  template<typename eT>
  void Initialize(arma::Cube<eT>& W,
                  const size_t rows,
                  const size_t cols,
                  const size_t slices)
  {
    W.set_size(rows, cols, slices);

    for (size_t i = 0; i < slices; i++)
      Initialize(W.slice(i), rows, cols);
  }
```

Reply:
Do you build with `DEBUG=ON` (compiler optimization off)?

Reply:
> @zoq Marcus, this might be interesting for you. Why didn't you write the function like the `arma::Mat` specialization

At the time I wrote the code I haven't thought about passing a non empty object. I think @mrityunjay-tripathi fix using `set_size` should work in both cases.

Reply:
Let me know if you need any help with the changes.